### ISSUE TITLE
HTML5 Boolean attributes always set

### ISSUE THREAD

**Issue by Stolzenhain** [javve/list.js#556]
… yes, we shouldn't complain now that there are massively efficient [custom attributes](http://listjs.com/examples/data-attributes-custom-attributes/)! 
Alas boolean attributes don't work because they are set once the attribute exists, which means using `valueNames` like

    {
      name: 'sticky',
      attr: 'checked'
    },

produces

    <input class="sticky" checked="" type="checkbox">

which even with an empty value still counts as a set attribute (eg. checkbox is rendered checked)

Is there

1. some way to not output empty attributes or
2. override/filter the attribute output function or
3. some other way I didn't get?

As stated in the title this concerns all boolean attribures like `disabled`, `slected`, `checked` and so on, which would be especially helpful for inline form elements.

**Comment by mcglonelevi** [comment]
I did some digging through the code, and I think calls to setAttribute need to be replaced with a custom version that looks for boolean attrs and calls removeAttribute as appropriate.

I found this article on stack overflow that outlines a few boolean attrs:
https://stackoverflow.com/questions/706384/boolean-html-attributes

There would need to be some significant testing on IE prior to pulling in as part of the code.  It looks like javve maintains a repo for getAttribute.  I think a separate repo for setAttribute and removeAttribute would be appropriate as well.

**Comment by mcglonelevi** [comment]
Also, this is a blocker on something I am working on at work.  Basically, we are using checkboxes, and having ListJS set the checked property by default is not acceptable for our use case.

**Comment by Stolzenhain** [comment]
Yup … this would need some sort of cleanup process that combs through empty attributes. But as you pointed out, this might have unforeseen consequences where you might want those (eg. CSS selectors containing `[attribute]`)
That being said; I have trouble recalling the exact circumstances of this; seeing this as potentially complicated, it should probably handled by some sort of hook doing this sort of cleanup on blacklisted HTML5 attribute names only to ensure validation.

**Comment by mcglonelevi** [comment]
I don't think CSS selectors containing `[attribute]` would be that difficult to get around.  There are many like:
```
:disabled
:checked
:optional
:enabled
```
That are used to select boolean attributes purely on whether or not a property exists.

The setAttribute logic should look something like this:
```
const blacklist = ['disabled', 'checked'];

function setAttribute(name, value) {
    if (blacklist.indexOf(name) === -1) {
       // run regular setAttr logic
    } else {
        // run logic to add/remove attr based on true/false
    }
}
```

Am I missing something?

**Comment by Stolzenhain** [comment]
@mcglonelevi : This looks good! Thinking this over, a less embracing approach (e.g. w/o new functions or overrides) could be patched in the [`setValue()` function](/javve/list.js/blob/master/src/templater.js#L101) like this – canceling the function if it is an empty boolean attribute:

```
// make sure HTML5 booleean attributes don't contain empty values
const blacklist = ['disabled', 'checked'];
if (
  (value.length < 1) && //empty value
  (blacklist.indexOf(name) > -1) //
)
  return;
```

--

> I don't think CSS selectors containing [attribute] would be that difficult to get around. There are many like:

What I was implying is that there might already be JS or CSS selector code that takes into account generated empty attributes (like `data-` attributes or empty `src` values) and could potentially break current sites if we generally removed all empty attributes (e.g. no blacklisting).
